### PR TITLE
Storage Queue Service

### PR DESCRIPTION
**Pull request by Jinming-Hu**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp:36]: still curious about this.  No-credential client?

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp:55]:
```suggestion
    Azure::Response<Models::ServiceStatisticsResult> GetStatistics(
```

File: sdk/storage/azure-storage-queues/src/queue_client.cpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-queues/src/queue_client.cpp:112]: Is there a reason why we need a protocol-model-options and a public-model-options?  
I would expect the protocol layer to always use the public-model-options.

File: sdk/storage/azure-storage-queues/src/queue_responses.cpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-queues/src/queue_responses.cpp:13]: the `CurrentPageToken` is missing to be updated

**Review by vhvb1989 — COMMENTED** [review]
- We require all public headers to be fully documented before merging. Please add Doxygen.
- We require a changelog

- Side Question:  Why do we use public-operation-options and protocol-operation-options?  Is it because of the auto-gen?  Are we paying an extra copy for the options on every operation?

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:1612]: This is auto-generated. What's the benefit of your suggested change?

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:1651]: This is auto-generated code. Maybe after we re-write the generator.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:35]: SAS auth

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: All other clients in other services have this method. Why shouldn't queue have it?

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:43]: I want to finish all public API interfaces and generate API view, so that we can get ready for API review.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp:36]: This is correct. No-credential client.

File: sdk/storage/azure-storage-queues/src/queue_client.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/src/queue_client.cpp:112]: Convenience layer API model and protocol layer API model was discussed like more than a year ago, and we landed on current model.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/src/queue_client.cpp:112]: Check the email titled _About the design on protocol layer_ from Jeffrey on Apr. 20, 2020.

Bring this up in next week's working group meeting if you want to overturn the conclusion.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:2040]: > If yes, would it help if you get the ref as const& ?

No, sometimes we need to extract binary body or xml body from the response. So it has to be non-const.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:2040]: > why? You can just use the pHttpResponse. Is there something you are winning here?

Not really. But is there something to win with `pHttpResponse` directly?

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:600]: This is auto-generated code. Generator is not smart enough to know if `options` will be used. So to get rid of compiler warning, I just added `(void)options` in each function.

**Comment by Jinming-Hu** [comment]
> * We require all public headers to be fully documented before merging. Please add Doxygen.

I want to get API view done first because according to our experience in blob/file services, API review usually takes longer time. I'll make up the docs in another PR.

> * We require a changelog

Ok

> * Side Question:  Why do we use public-operation-options and protocol-operation-options?  Is it because of the auto-gen?  Are we paying an extra copy for the options on every operation?

This is agreed by the working group. Check the email titled _About the design on protocol layer_ from Jeffrey on Apr. 20, 2020.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp]: I'll resolve all comments about auto-gen

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:1606]: I'll resolve all comments about auto-gen

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:1644]: I'll resolve all comments about auto-gen

File: sdk/storage/azure-storage-queues/src/queue_responses.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/src/queue_responses.cpp:13]: `CurrentPageToken` and `NextPageToken` are updated in the call of `ListQueues()`

File: sdk/storage/azure-storage-queues/src/queue_client.cpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/src/queue_client.cpp:112]: The rationale behind this design is that it requires human intervention to decide which parameters are mandatory and which are optional. But the input of auto-gen (which is usually swagger) doesn't contain this information. So there's no way auto-gen can know which parameters should be in parameter list and which should be in options bag.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:64]: GetXMethod should return X rather than GetXResult. This is agreed by C++ working group. Ref to the email titled _[SDK model type naming consistency] Client library GetX Methods should return model types named X and not GetXResult_ by Ahson.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:78]: GetXMethod should return X rather than GetXResult. This is agreed by C++ working group. Ref to the email titled _[SDK model type naming consistency] Client library GetX Methods should return model types named X and not GetXResult_ by Ahson.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_service_client.hpp:55]: GetXMethod should return X rather than GetXResult. This is agreed by C++ working group. Ref to the email titled _[SDK model type naming consistency] Client library GetX Methods should return model types named X and not GetXResult_ by Ahson.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:82]: I feel `ReceiveX()` is very like `GetX()`. So I feel for this one we should follow the rule that `GetX()` returns `X`.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:55]: GetXMethod should return X rather than GetXResult. This is agreed by C++ working group. Ref to the email titled _[SDK model type naming consistency] Client library GetX Methods should return model types named X and not GetXResult_ by Ahson.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:2024]: You would remove this line of code, and the very small memory allocation in the stack for the created reference :)

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp:1606]: If it is auto-gen, I don't care. It's fine :)

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: That's fine then, for consistency with the storage clients. 
Question is, should clients be exposing this at all?  (for all services / languages? )  @JeffreyRichter

Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: And, shouldn't it be a const& returned here?  This gives the impression that the Url can change at some point in the future, so it needs to be generated every-time for the customer

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:43]: I totally understand.  However, my understanding is that we want to avoid having imcompleted/no-ready-to-ship code in `main` branch at any given time. That's the reason behind the PR-template telling all the PR-requeriments.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:78]: This is not a `GetXMethod` is it?  This is a `ReceiveXMethod`  .  Would the rule cover all `verbXMethod` like
- DeleteX
- PutX
- SendX
- LeaseX
- etcX
??

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (JeffreyRichter) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: Yes, all XxxClient types should have a GetURL method on them. This is required to get the URL and pass it to another client (with different pipeline options or credentials) or pass it to another machine so it operate on the resource.

File: sdk/storage/azure-storage-queues/src/queue_client.cpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/src/queue_client.cpp:112]: Since the protocol layer is implementation detail, I would make all methods to take only const& to all values directly, without introducing new models. That would save the extra-copy.  And the auto-gen would just flatten the models into a full list of input parameters for the method.
Anyway, This is not something I would ask to be changed at this point. It's late for that. I can live with this. Thank you for explaining it.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: https://github.com/Azure/azure-sdk-for-cpp/issues/2701

Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:37]: Make the url string a private field of the clients and return a const& to it instead of creating a new copy everytime

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:43]: Talked to the team about this.
The agreement and expectation is to avoid merging/pushing public-APIs without a well-written documentation. So, please add documentation for all the public exposed files before merging

**Review by vhvb1989 — COMMENTED** [review]
Missing Doxygen

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (katmsft) [inline sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:86]: It is not consistent to have `PeekedQueueMessage` from `PeekMessage` + `QueueMessage` from `ReceiveMessage`.
If we are honoring the rule at all, it should also be `QueueMessage` from `PeekMessage` or `PeekMessageResult`.
Or `ReceivedQueueMessage` from `ReceiveMessage`, which I personally think is too much.

**Review by katmsft — APPROVED** [review]

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/protocol/queue_rest_client.hpp]: updated our generator https://github.com/Azure/azure-sdk-for-cpp/pull/2703

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:78]: discussing via email

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:82]: discussing via email

**Review by vhvb1989 — COMMENTED** [review]
Doxygen for public layer still missing. Won't approve it without it.

**Comment by Jinming-Hu** [comment]
> Doxygen for public layer still missing. Won't approve it without it.

Working on it. Don't rush me.😶

**Comment by Jinming-Hu** [comment]
TODO: Add test cases for account sas

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_credential.hpp
Comment (vhvb1989) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/storage_credential.hpp:66]: Make sure to set VCPKG config file to depend on the latest storage common package containing this new structs.

File: sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp
Comment (vhvb1989) [reply sdk/storage/azure-storage-queues/inc/azure/storage/queues/queue_client.hpp:3]: What happens when we don't add file-doxygen is that we let the files-documentation empty like:

image

We want the files section to tell customers also meaningful info like:

image

**Review by vhvb1989 — COMMENTED** [review]
Can you just add file-docs ?  I think that's my only last request :)

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_credential.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_credential.hpp:66]: Specifying version in vcpkg.json doesn't work very well now, and it's removed from other packages.

Comment (antkmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_credential.hpp:66]: We still can and should request minimum versions via `find_dependency()` and `find_component()`, search for other occurrences in our codebase - they all are versioned now.

I think that version constraints may come back as early as before the next release - https://github.com/Azure/azure-sdk-for-cpp/issues/2735 has details.

And then, having the min-versions in `find_dependency/component()` will be very useful to re-add `version>=` in the json manifests.

**Review by antkmsft — COMMENTED** [review]
I think the `azure/storage/queues.hpp` does not include all files as it should.

Ideally, there is also a corresponding unit test that makes sure that that header does indeed include all that it should, you can see example in core or identity.

**Pull request merged**
